### PR TITLE
Hamiltonian cycle added

### LINKED ISSUE #820: Hamiltonian Cycle
Add an algorithm to find Hamiltonian Cycle, in the following languages:
- C/C++
- Python
- Java

### PR DESCRIPTION
**Pull request by KanchanThareja**
Fixes #820

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Sample Input / Output.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

#### Changes proposed in this pull request:

-  implemented hamiltonian cycle

#### Languages Used:
- c++
- python
- java

#### Files Added:
- Hamiltonian_Cycle.cpp
- Hamiltonian_Cycle.java
- Hamiltonian_Cycle.py

### CODE REVIEW

File: Hamiltonian_Cycle/Hamiltonian_Cycle.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    } 
+
+    // Function to solve hamiltonian cycle problem
+    boolean helper(int graph[][], int path[], int index) 
+    { 
+        if (index == vertices) 
+        { 
+            if (graph[path[index - 1]][path[0]] == 1) 
+                return true; 
+            else return false;
```
Comment (somya-kapoor) [inline Hamiltonian_Cycle/Hamiltonian_Cycle.java]: Please fix the indentation ( write return false in a new line).

**Comment by MastersAbh** [comment]
The work is good @KanchanThareja ,
but while going through the code, I found a lot of similarity with the code available at [https://www.geeksforgeeks.org/hamiltonian-cycle-backtracking-6/] . 
Only the function names are changed in addition to minor omission of print function.
Even the example used is same as in the above url.

Please re-code it and then it would be good to review.
@jainaman224

**Comment by KanchanThareja** [comment]
@MastersAbh I do get your point since i learnt the algorithm from there itself
But i implemented it on my own self. It is obvious the logic would be same and also that i took example from there. I dont think i can change the code entirely.

**Comment by MastersAbh** [comment]
I agree, but only changing the variable names is not called coded by own.
Also, you have made the same mistake made on the GFG link I shared, where the array of graph is taken as bool datatype, which can easily be taken as a vector or an array normally. So, I think you should focus on writing it from scratch without taking inputs from here.

**Comment by KanchanThareja** [comment]
@MastersAbh You are free to choose any data structure you wish to. Its not a mistake.

**Comment by MastersAbh** [comment]
No, @KanchanThareja  It is implemented as a vector there. I agree for the problem statement, it won't matter because there's no use of that. But, the graph variable is implemented as a vector/array, so it should be that.

**Comment by MastersAbh** [comment]
And I don't want you to change the example, you can do a lot of changes if you have got the idea behind Hamiltonian Cycle. Just changing variable names and examples would not help me or you or anyone and it conflicts with the main intention behind the project.

**Comment by KanchanThareja** [comment]
it is a boolean 2D matrix. Whats a problem in that?

**Comment by MastersAbh** [comment]
Correct, there is no problem in that for this question. 
I was just pointing out to the similarity level of your code. It could have been implemented differently.

**Comment by MastersAbh** [comment]
Please instead of extending this conversation, lets focus on what I'm trying to say and try to implement it on your own.

**Comment by KanchanThareja** [comment]
Then why are you pointing out vector? I don't get your point.

**Comment by MastersAbh** [comment]
I am not pointing out one single point. I am just saying your code is way too similar and you've just changed variable names and it doesn't go along with the intention of the project, so you should try to implement it on your own.
There are many ways one can implement it if you have the idea behind the problem statement, in this case, Hamiltonian Cycles.

**Comment by KanchanThareja** [comment]
Alright if you think there is too much similarity i ll make changes though i still dont get your inputs.

**Comment by MastersAbh** [comment]
Sure, that would be great. 
Now, lets get to the point you have a doubt about.
I was suggesting one way of implementation change could have been using a vector of int, instead of 2d boolean matrix for graph, which could have opened scope for more implementation changes.
Let us get over it now.

**Comment by KanchanThareja** [comment]
Okay i changed it that way take a look

**Comment by MastersAbh** [comment]
Hey @KanchanThareja , 
It is not about making one change or changing variable names. It is about how well can you code what you have understood. I was quoting one thing that you can do in your implementation and not that you should change here.
Have a look at your code and the code available at GFG, don't you find many similarities?
Please try to code afresh, else it would be flagged as plagiarism. It won't look different if you code fresh and if you have the idea of hamiltonian cycles.

**Pull request closed without merging**